### ISSUE TITLE
Missing remote object

### ISSUE THREAD

**Issue by paul321** [electron/electron#8205]
I have a function that is called several times in a loop that is trying to access cookies, but after few times i get the following error:

Electron: 1.4.11

Cannot get property 'cookies' on missing remote object 16127
Error: Cannot get property 'cookies' on missing remote object 16127
    at throwRPCError (D:\devel\work\js\microstockr.desktop\node_modules\electron-prebuilt\dist\resources\electron.asar\browser\rpc-server.js:143:17)
    at EventEmitter.<anonymous> (D:\devel\work\js\microstockr.desktop\node_modules\electron-prebuilt\dist\resources\electron.asar\browser\rpc-server.js:368:7)
    at emitThree (events.js:116:13)
    at EventEmitter.emit (events.js:194:7)
    at WebContents.<anonymous> (D:\devel\work\js\microstockr.desktop\node_modules\electron-prebuilt\dist\resources\electron.asar\browser\api\web-contents.js:231:13)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)

The code is:
var session = require('electron').remote.session;
session.defaultSession.cookies.get({ url: url }, function(err, domainCookies) {.....});

**Comment by MarshallOfSound** [comment]
@paul321 I was under the impression this issue had been fixed in `1.4.11` can you verify you are indeed on that version (clean install to double check).

Also I am unable to reproduce on `1.4.11`.  My testing code is this.

```js
setInterval(() => {
  var session = require('electron').remote.session;
  session.defaultSession.cookies.get({ url: 'https://ffo.com' }, function(err, domainCookies) {});
  console.info('run')
}, 1)
```

Which I let run almost 100,000 times with no errors.  Can you provide a reproducible example?

**Comment by paul321** [comment]
It's not happening all the time and it's random, sometime is the cookies that cannot be found sometimes other remote object, i'm trying to figure out when it does, I'm on 1.4.11

**Comment by AoDev** [comment]
Hi. I'd like to say that I have the same error with a different process.
In my case.  `Cannot get property format on missing remote object`. 

I am using: 
```
"electron": "1.4.13",
"electron-builder": "10.15.1",
```

It is also "random".
That's the reason why I can't paste the error stacktrace now. I have relaunched the process multiple times but, obviously, it had to work fine for now. I will try to reproduce it later and give more info.

**Comment by AoDev** [comment]
Ok, finally I could reproduce it. I have no idea why it "failed" this time.

Here is the stacktrace:

```
Uncaught Error: Cannot get property 'format' on missing remote object 246438
Error: Cannot get property 'format' on missing remote object 246438
    at throwRPCError (/Applications/Kaiku.app/Contents/Resources/electron.asar/browser/rpc-server.js:143:17)
    at EventEmitter.<anonymous> (/Applications/Kaiku.app/Contents/Resources/electron.asar/browser/rpc-server.js:368:7)
    at emitThree (events.js:116:13)
    at EventEmitter.emit (events.js:194:7)
    at WebContents.<anonymous> (/Applications/Kaiku.app/Contents/Resources/electron.asar/browser/api/web-contents.js:231:13)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)
```

Here is the code that is probably failing but I can't be 100% sure.
Note, I was rewriting my app, some of this code is still legacy. This didn't happen before.

Basically, I call a library method that:

- should scan the filesystem
- push a task in a queue for each file found
- run the callback for each file, one after another

This is a code running in the renderer process:
_library.js_
```
const scanner = remote.require('./lib/libraryScanner')
...
scanner.scan(...)
...
 // save data in webSQL
  function saveToDb () {
    coverFormat = metadata.picture.length ? metadata.picture[0].format : null
    ...
  }
...
```
What I think is failing, per the error message: `metadata.picture[0].format`

_libraryScanner.js_
This particular bit, I am not too sure how Electron handles it. If it runs on the main process because I call it with `remote.require('./lib/libraryScanner')` as I assumed. But maybe I am doing something wrong.

```
var fs = require('fs')
var mm = require('musicmetadata')
var filewalker = require('filewalker')
var async = require('async')

function scan (folder, onMetadata, onComplete) {

  function processFile (filePath, qNext) {
    const stream = fs.createReadStream(filePath)
    mm(stream, function (err, metadata) {
      if (!err) {
        metadata.filePath = filePath
        stream.close()
        onMetadata(metadata, function () {
          qNext()
        })
      }
      else {
        qNext()
      }
    })
  }

  /**
   * Use an asynchronous queue to handle the data process
   * @see  https://github.com/caolan/async#queue
   */
  const q = async.queue(processFile, 1)

  function addToQueue (p, s, fullPath) {
    q.push(fullPath)
  }

  function handleError (err) {
    console.log(err)
    onComplete(err)
  }

  function onAllFilesFound () {
    q.drain = onComplete
    found = this.files
  }

  filewalker(folder)
    .on('file', addToQueue)
    .on('error', handleError)
    .on('done', onAllFilesFound)
    .walk()
}
```

**Comment by chance-an** [comment]
I also encountered this issue while interacting with some file related APIs and `musicmetadata`. I could work around it by [deepcopying](https://www.npmjs.com/package/deepcopy) it after received the data in the renderer process. I also noticed significant performance boost after deepcopying.

Of course it might not work if you need to use the method/function calls on the received data object.

**Comment by HitomiTenshi** [comment]
I'm struggling with the same issue here. It happens randomly and I can't exactly reproduce it every time. It just happens sometimes:

```
zone.min.js:1 Uncaught Error: Error in file:///C:/Users/johann.rekowski/Documents/Apps/app/src/app/components/home/home.component.html:25:17 caused by: Cannot call function 'isMaximized' on missing remote object 3
Error: Cannot call function 'isMaximized' on missing remote object 3
    at throwRPCError (C:\Users\johann.rekowski\Documents\Apps\app\node_modules\electron\dist\resources\electron.asar\browser\rpc-server.js:143:17)
    at EventEmitter.<anonymous> (C:\Users\johann.rekowski\Documents\Apps\app\node_modules\electron\dist\resources\electron.asar\browser\rpc-server.js:339:7)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
    at WebContents.<anonymous> (C:\Users\johann.rekowski\Documents\Apps\app\node_modules\electron\dist\resources\electron.asar\browser\api\web-contents.js:231:13)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)
```

**Comment by neurosnap** [comment]
I'm seeing this issue as well, any updates on this?

**Comment by kevinsawicki** [comment]
Is anyone able to provide a code sample that reproduces this? Would be happy to take a look but I've been unable to reproduce this since the the last fix for it.

**Comment by HitomiTenshi** [comment]
I fixed my issue by using `ipcMain` and `ipcRenderer`. Using the remote object directly from the renderer thread seems to cause some bugs.

**Comment by kevinsawicki** [comment]
> Using the remote object directly from the renderer thread seems to cause some bugs.

Yes, that is true, the `missing remote object` exception only occurs for objects, properties, methods accessed via the `remote` API.

Until the `remote` issue is resolved, switching code from using `remote`  to `ipcMain`/`ipcRenderer` will prevent the exception from being thrown.

**Comment by ccnokes** [comment]
@kevinsawicki what is the `remote` issue?

**Comment by kevinsawicki** [comment]
>  what is the remote issue?

The issue is that remote objects are somehow being incorrectly removed from the objects registry in the main process even though they are still referenced in the renderer process.

The stack traces people are seeing in this issue is what happens when this problem occurs, a remote object is accessed but the main process does not have it in its object registry so it is unable to call or access the requested method or property and so it throws this missing object error.

**Comment by ccnokes** [comment]
Yeah, that's exactly what's going on for me. Is there an open issue/in-progress PR somewhere for that that I can track? (Or maybe this is it?) 

Let me know if there's any additional info I can provide to help or if there any workarounds for it 😸

**Comment by kevinsawicki** [comment]
> Or maybe this is it?

This is the issue to track this error down.

> Let me know if there's any additional info I can provide to help or if there any workarounds for it

Are able to reproduce it (even if only 1 out of X number of times) with a smallish code example that you could provide here?

**Comment by ccnokes** [comment]
> smallish code example

I'll see what I can do... it might be a few days before I can get back to you on that. Like others, when the error happens, it seems to be sort of random. I.e. its working just fine and then all of the sudden I get the `missing remote object` error. I'll try to track it down though. Thanks!

**Comment by AoDev** [comment]
My 2 cents, it could be garbage collector related. I have changed my process to make it much less memory intensive and the error does not happen anymore. That's also why it can be random, garbage collection may be triggered at any time. But I have nothing to prove this theory right now.

**Comment by paul321** [comment]
We got this issue also when calling the function that is setting the badge icon from code

Uncaught (in promise) Error: Could not call remote function 'setOverlayIcon'. Check that the function signature is correct. Underlying error: Error processing argument at index 1, conversion failure from 6
Error: Could not call remote function 'setOverlayIcon'. Check that the function signature is correct. Underlying error: Error processing argument at index 1, conversion failure from 6

If we run the function from dev console like this App.setBadge('1') it's working find, called from other functions doesn't work

**Comment by BrandonSmith** [comment]
FWIW... I cannot reproduce locally, but my error catcher for a deployed app in the field catches the following...

`Cannot call function 'setBadgeCount' on missing remote object`

...on the call...

`remote.app.setBadgeCount(count)`

I can go the `ipc` route, but it seems odd that the `app` property goes missing.

**Comment by thomasmeadows** [comment]
same issue here, kinda random at app boot time

**Comment by ccnokes** [comment]
Just reporting back here, I've never been able to reproduce this one. However, I did discover a memory leak in my app caused by <webview> DOM getting destroyed before I had a chance to clean up some resources associated with it, i.e. `this.webContents.session.webRequest.onErrorOccurred(null)`, etc. It seems that if the DOM for a webview is destroyed then you lose your handle on it's webContents, etc so you the above call errors out, and you get a leak. 

So basically, being extra diligent in checking for leaks helped me solve this issue. It's easy to leak things with event emitters (a lot of Electron APIs are event emitters) especially. Hopefully this helps someone.

**Comment by zcbenz** [comment]
I believe we have already a few bugs may cause this issue in recent versions, I'm closing this issue now but if anyone still sees in v1.7.x, please report back and I'll reopen.

**Comment by magne4000** [comment]
Using `1.7.2` here and still randomly having this error
```
Error: Uncaught Error: Cannot get property 'id' on missing remote object 3637
Error: Cannot get property 'id' on missing remote object 3637
```

**Comment by WriterDuet** [comment]
I'm seeing this randomly on 1.7.4

**Comment by itshaadi** [comment]
same problem happens randomly on 1.6.11 (windows 7)

> Uncaught Error: Cannot get property 'type' on missing remote object 1879
Error: Cannot get property 'type' on missing remote object 1879
    at throwRPCError (D:\dev\st\node_modules\electron\dist\resources\electron.asar\browser\rpc-server.js:143:17)
    at EventEmitter.<anonymous> (D:\dev\st\node_modules\electron\dist\resources\electron.asar\browser\rpc-server.js:384:7)
    at emitThree (events.js:116:13)
    at EventEmitter.emit (events.js:194:7)
    at WebContents.<anonymous> (D:\dev\st\node_modules\electron\dist\resources\electron.asar\browser\api\web-contents.js:256:37)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)
    at throwRPCError (D:\dev\st\node_modules\electron\dist\resources\electron.asar\browser\rpc-server.js:143:17)
    at EventEmitter.<anonymous> (D:\dev\st\node_modules\electron\dist\resources\electron.asar\browser\rpc-server.js:384:7)
    at emitThree (events.js:116:13)
    at EventEmitter.emit (events.js:194:7)
    at WebContents.<anonymous> (D:\dev\st\node_modules\electron\dist\resources\electron.asar\browser\api\web-contents.js:256:37)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)
    at metaToValue (D:\dev\st\node_modules\electron\dist\resources\electron.asar\renderer\api\remote.js:234:13)
    at Object.descriptor.get [as type] (D:\dev\st\node_modules\electron\dist\resources\electron.asar\renderer\api\remote.js:136:16)
    at _traverse (D:\dev\st\node_modules\vue\dist\vue.runtime.common.js:3053:32)
    at _traverse (D:\dev\st\node_modules\vue\dist\vue.runtime.common.js:3049:19)
    at _traverse (D:\dev\st\node_modules\vue\dist\vue.runtime.common.js:3053:19)
    at _traverse (D:\dev\st\node_modules\vue\dist\vue.runtime.common.js:3049:19)
    at _traverse (D:\dev\st\node_modules\vue\dist\vue.runtime.common.js:3053:19)
    at _traverse (D:\dev\st\node_modules\vue\dist\vue.runtime.common.js:3049:19)
    at _traverse (D:\dev\st\node_modules\vue\dist\vue.runtime.common.js:3053:19)
    at _traverse (D:\dev\st\node_modules\vue\dist\vue.runtime.common.js:3053:19)
>

**Comment by abnerlee** [comment]
`v1.7.2`

```
ERROR 9/4/2017, 9:15:45 PM  [RenderProcess 1][Error] Error: Cannot call function 'getItem' on missing remote object 110
Error: Cannot call function 'getItem' on missing remote object 110
    at throwRPCError (/Users/abner/workplace/HybirdApp/TypeMark/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:143:17)
    at EventEmitter.<anonymous> (/Users/abner/workplace/HybirdApp/TypeMark/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:354:7)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:204:7)
    at WebContents.<anonymous> (/Users/abner/workplace/HybirdApp/TypeMark/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/api/web-contents.js:256:13)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:194:7)
    at metaToValue (/Users/abner/workplace/HybirdApp/TypeMark/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/renderer/api/remote.js:234:13)
    at Object.remoteMemberFunction (/Users/abner/workplace/HybirdApp/TypeMark/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/renderer/api/remote.js:118:18)
    at Object.doRefreshImageCopyStatus (file:///Users/abner/workplace/HybirdApp/TypeMark//app/window/MenuHelper.js:216:50)
    at file:///Users/abner/workplace/HybirdApp/TypeMark//app/document/File.js:1075:32
```

**Comment by JohnWeisz** [comment]
For me, this occurred today, in `1.7.9`.

**Comment by deerares** [comment]
The problem is still alive. Is there at least any plans to fix this?

**Comment by thomasmeadows** [comment]
@zcbenz can you reopen this?

**Comment by Thomas101** [comment]
We've seen this come up a few times and always seems to revolve around new windows opening and closing with `nativeWindowOpen` enabled.

I'll keep trying to get a simple test-case together, but as other have mentioned it's sporadic.

**Comment by funwithtriangles** [comment]
Seeing this with `1.8.2-beta.2`. Can confirm that it seems to be something to do with new windows and I am using the `nativeWindowOpen` setting.

**Comment by jfurneaux** [comment]
Hi everyone, did you figure this out? We're stuck on this and keen to hear how people have handled. @funwithtriangles @Thomas101 @thomasmeadows @deerares

**Comment by joneian** [comment]
I'm seeing this problem being reported for my App too. Is a work around to keep a reference to the remote objects in the render process to stop them being garbage collected? (or is the problem that they're getting collected in the main process?)

**Comment by nicbarker** [comment]
I'm currently experiencing this in 1.7.8 on Mac OS X, it happens sporadically and only with webkit dev server (i.e. having a window open for a long time with a bunch of refreshes)

**Comment by sethlu** [comment]
I encountered this issue just yesterday, and after spending some time investigating, I think it's due to objects at times being released when they are not supposed to. I can post a fix for the "missing object" but I'll try to find the root cause of this later today.

**Comment by damoqiongqiu** [comment]
Finally, I think I have found the reason for this issue.

If we are using ipcRenderer inside a Renderer thread, including a webview tag.

When the renderer navigated to another URL, some callbacks like setTimeout of setInterval may invoke the methods on that renderer or webview instance after a period of time.

This will definitely cause "missing remote object ..."

So, the problem here is, how can we prevent ourselves from using ipcRenderer in some async callbacks?

**Comment by damoqiongqiu** [comment]
After several days of testing, I found this issue is not existing in V 1.8.x, but reappears in V 2.x.

**Comment by zcbenz** [comment]
@damoqiongqiu Can you write a minimal app that I can use to reproduce the problem?

**Comment by mailtokartik1** [comment]
@zcbenz https://github.com/electron/electron/issues/13137
Here is a repository you can use to reproduce the bug

**Comment by mailtokartik1** [comment]
@sethlu are you close to finding a workaround or a fix?

**Comment by sethlu** [comment]
@mailtokartik1 I’ll look into this a little more over the weekend. A couple of weeks ago I had a weird workaround but hasn’t tested it on the repro just yet.

**Comment by mailtokartik1** [comment]
@sethlu right! I have been able to run my code successfully in v2.0.0 though. This PR was the cause of the error after I researched #12342. I reverted back this PR in my v2.0.0 and the missing remote error went away. Have a look into this too.

**Comment by sethlu** [comment]
After looking into this issue a little more today I believe it boils down to the timing for garbage collection in the renderer process interfering with an assumption we made at the time the `remote` module was written. Currently the issue should reliably appear when one tries to quickly access some `remote` objects that are `const` on the browser process side, e.g. web contents or cookies.

For https://github.com/electron/electron/issues/13137, I think commit https://github.com/electron/electron/pull/12342/commits/949b92030b015caedcaefe8fdb239154cfcc6c6c doesn't directly cause this reoccurring issue but can come into play... Here are some logs collected (& annotated) to illustrate the problem based on https://github.com/electron/electron/issues/13137:

```sh
get webcontents                            # Message ELECTRON_BROWSER_CURRENT_WEB_CONTENTS received, from `webPreferences = remote.getCurrentWebContents().getLastWebPreferences()`
webcontents 17179869185 retain 1 counts 1
get 1
webcontents 17179869185 retain 2 counts 1
get webcontents                            # Message ELECTRON_BROWSER_CURRENT_WEB_CONTENTS received, from `webPreferences = remote.getCurrentWebContents().getLastWebPreferences()`
webcontents 25769803777 retain 2 counts 2
get 2
get 2
get 2
get webcontents                            # Message ELECTRON_BROWSER_CURRENT_WEB_CONTENTS received, from `webPreferences = remote.getCurrentWebContents().getLastWebPreferences()`
webcontents 25769803777 retain 2 counts 2  # Our objects-registry counts 2 because webcontents 25769803777 has already retained this obj
get 2
get 2
get 2
gc                                         # Garbage collection initiated from renderer process to destruct obj 1 in main process
webcontents 17179869185 remove 1
release 1 counts 0
gc                                         # Garbage collection initiated from renderer process to destruct obj 2 in main process
gc                                         # Garbage collection initiated from renderer process to destruct obj 2 in main process
webcontents 25769803777 remove 2           # Obj 2's reference count decrements to 1
release 2 counts 1
webcontents 25769803777 remove 2           # Obj 2's reference count decrements to 0 (currently we're not checking if 25769803777 really owns 2 before decrementing reference count)
release 2 counts 0
                                           # When webcontents 17179869185 tries to get obj 2 the exception is thrown
```

Say the garbage collection kicks in earlier for `webcontents 25769803777 remove 2`, the problem will then disappear since the reference counting adds up. Illustrated below:

```sh
get webcontents                            # Message ELECTRON_BROWSER_CURRENT_WEB_CONTENTS received, from `webPreferences = remote.getCurrentWebContents().getLastWebPreferences()`
webcontents 17179869185 retain 1 counts 1
get 1
webcontents 17179869185 retain 2 counts 1
get webcontents                            # Message ELECTRON_BROWSER_CURRENT_WEB_CONTENTS received, from `webPreferences = remote.getCurrentWebContents().getLastWebPreferences()`
webcontents 25769803777 retain 2 counts 2
get 2
get 2
get 2
gc                                         # Garbage collection initiated from renderer process to destruct obj 2 in main process
webcontents 25769803777 remove 2           # Obj 2's reference count decrements to 1
release 2 counts 1
get webcontents                            # Message ELECTRON_BROWSER_CURRENT_WEB_CONTENTS received, from `webPreferences = remote.getCurrentWebContents().getLastWebPreferences()`
webcontents 25769803777 retain 2 counts 2
get 2
get 2
get 2
gc                                         # Garbage collection initiated from renderer process to destruct obj 1 in main process
webcontents 17179869185 remove 1
release 1 counts 0
gc                                         # Garbage collection initiated from renderer process to destruct obj 2 in main process
webcontents 25769803777 remove 2           # Obj 2's reference count decrements to 1
release 2 counts 1
                                           # When webcontents 17179869185 tries to get obj 2, it's still available
```

Now to resolve the issue, here are a few approaches:

- I think `remote` object assumes that the objects returned are unique to each other (so attaching an hidden id to an object works well); however, when returning an const object from the browser process, it may end up correlating to multiple proxying objects from one renderer process, leading to multiple releases of the same object -- to resolve this we may try to return a same proxying object for a same object from the main process so the remote object is only freed once, rather than proportional to the number of proxies

Not so good ones?! Sound like workarounds...

- Always increment reference counts -- a web contents may end up accumulating multiple reference counts for a same remote object (which won't lead to the missing remote object issue)
- Before decrementing the reference count for an object, check if it's owned by the claimed web contents

**Comment by MarshallOfSound** [comment]
> Always increment reference counts -- a web contents may end up accumulating multiple reference counts for a same remote object (which won't lead to the missing remote object issue)

Another option here is to reuse the proxy for the same remote object

**Comment by sethlu** [comment]
@MarshallOfSound on page navigations the preload script wouldn't be able to retain the same proxy for the same remote object -- I believe that's what leads to twice garbage collections initiated by renderer processes on a same remote object.

**Comment by sethlu** [comment]
Some status update: I think this might be a duplicate of https://github.com/electron/electron/issues/9387, which root cause didn't get addressed in https://github.com/electron/electron/pull/9389.

@tarruda traced the sequence of events here https://github.com/electron/electron/issues/9387#issuecomment-299606906:

> - initial load
> - hello.js was added to objects-registry and referenced count increased
> - reload
> - hello.js was added to objects-registry(no reference count increased)
> - after a while, garbage collection in renderer process caused the first reference to hello.js to be released
> - browser process removes hello.js from object-registry even though it is still being referenced by the second context.

... which pretty much overlaps over our scenario in https://github.com/electron/electron/issues/13137. Then the fix introduced in https://github.com/electron/electron/pull/9389 may result in the following sequence of events?

> - initial load
> - hello.js was added to objects-registry and referenced count increased
> - reload **& clear reference count in objects-registry**
> - hello.js was added to objects-registry ~(no reference count increased)~ **and referenced count increased**
> - after a while, garbage collection in renderer process caused the first reference to hello.js to be released
> - ~browser process removes hello.js from object-registry even though it is still being referenced by the second context.~

The provided tests (https://github.com/electron/electron/pull/9389/commits/5654ff0d430b2902aedef4607094e92d4898a39c) couldn't surface the original issue: After running some tests I found that it turns out the `Hello` constructor function is not garbage collected (while the other objects are) so calling the constructor works just fine (ref: https://github.com/electron/electron/pull/9389/files#diff-24a81d8410a438c171cacad3e3468475R36). Our issue appears to haunt our remote `BrowserWindow`, required remote node modules and more.

---

I'll post a small proposal for some small change on remote ref counting in a little bit.

**Comment by zcbenz** [comment]
@sethlu As far as I understand, the root cause seems to be: after page navigation, garbage collection still happens for the objects in the old page?

When designing the original `remote` module, we had the assumption that every page navigation should restart the renderer process, but recent changes have introduced new features that can break this assumption.

To fix this issue, I think we can make the `remote` module clear references for current __page__, instead of current __WebContents__, e.g. something like this:

```js
// rpc-server.js
ipcMain.on('ELECTRON_BROWSER_CONTEXT_RELEASE', (e, pageId) => {
  objectsRegistry.clear(pageId)
  e.returnValue = null
})
```

```js
// remote.js
process.on('exit', () => {
  ipcRenderer.sendSync('ELECTRON_BROWSER_CONTEXT_RELEASE', currentPageId)
})
```

**Comment by sethlu** [comment]
> As far as I understand, the root cause seems to be: after page navigation, garbage collection still happens for the objects in the old page?

@zcbenz Thanks for the succinct summary! If we keep a different registry for each page the problem should be resolved. I'm not sure what's a good way to identify different pages for a webcontents?

**Comment by sethlu** [comment]
I'll drop this note here as well: If we can keep a different registry for each page (as what's brought up above), another problem which can happen is when the main browser window refers to the remote web contents object of a web view (or browser view), when the web view (or browser view) navigates and the garbage collection starts late, the reference count still gets decremented twice due to (1) clearing web contents and (2) garbage collection. This case should be fixed by reverting https://github.com/electron/electron/pull/9389/commits/6b5bd3b6ceb9457b091171bbde40700be71c66dd?

**Comment by zcbenz** [comment]
We have code in `objects-registry.js` listening to the `render-view-deleted` event of `WebContents` which is emitted when the renderer process is destroyed. It used to be the only way of clearing references of a page, but it does not emit when a page is reloaded without restarting the renderer process.

**Comment by sethlu** [comment]
@zcbenz I just ran some tests and it seems that right now our delete listener in `objects-registry.js` gets called twice on the same web contents because before page reload/navigation the owner is cleared in the object registry, on the second page load the listener will be attached again. So for our current set up we may are calling `clear` on the object registry for a web contents three times (once from `ELECTRON_BROWSER_CONTEXT_RELEASE` before page reload/navigation and twice around the time renderer process exits because the listener is attached twice).

**Comment by zcbenz** [comment]
@sethlu I also noticed that, in theory the event should emit immediately when page does navigation, but somehow Chromium is delaying all the events to when renderer process is destroyed. I think it should be a bug in Chromium, I'll guard that situation in the fix.

**Comment by sethlu** [comment]
@zcbenz A random follow-up thought: Right now the event `render-view-deleted` gets fired `N`-times on `N` page navigations/loads. I'm wondering the implications this observation has on the possible memory leak in web contents on previously loaded pages? 🤔 JavaScript callbacks from previous pages seem always waiting in the background.

**Comment by zcbenz** [comment]
@sethlu It seems that the `RenderView`s are not released when a page navigation happens, they are only released when the renderer process gets destroyed. We probably have something preventing the `RenderView` to be released.

**Comment by mdtanrikulu** [comment]
Electron v3.1.2 
OS: MacOS
Node: v10.6.0

``` js
Uncaught Error: Cannot call function 'onDidChange' on missing remote object 11
Error: Cannot call function 'onDidChange' on missing remote object 11
```

It happens in first run. After refresh the page everything looks okay.

**Comment by efonzie** [comment]
Electron V4.0.6

I know I'm a bit late to the party on this one, but I was recently running into the same issue described in this thread.  

Specifically, I was trying to export several PDFs using the browserWindow.webContents.printToPDF() method in a loop.  Once every 2-3 attempts at exporting 30+ pdfs I would have a single pdf fail to export and throw an error stating "cannot call 'printToPdf' on missing remote object". 

As I was unable to find a 'good' solution for this, I ended up simply wrapping the printToPDF() call in a try/catch clause and retrying if it failed.  This worked surprisingly well and I have yet to see the error show up again even though I'm currently only retrying a single time.  If necessary, you could set up your application to retry a set number of times before giving up.

I know its not an ideal solution, but if you're like me and needed some added reliability, this did the trick.
